File: examples/continuation_object_sweep_record.zig
//! Fixed, pointer-free evidence record for one committed object sweep.

const std = @import("std");
const core = @import("core");
const capsule = core.continuation_capsule;
const object_store = core.continuation_object_store;
const sweep = core.continuation_object_sweep;
const record = core.continuation_object_sweep_record;
const sweep_writer = core.continuation_object_sweep_writer;

pub fn main() !void {
    const input = demoInput(0x6a, 0x6b);
    var storage: [record.encoded_bytes]u8 = undefined;
    const encoded = try record.encodeV1(input, &storage);
    const decoded = try record.decodeV1(encoded);
    const expected = try record.expectationV1(input, decoded.record_sha256);
    _ = try record.decodeAndVerifyV1(encoded, expected);
    const plan = try record.appendPlanV1(encoded);

    var second_input = demoInput(0x6b, 0x6c);
    second_input.sequence = 2;
    second_input.previous_record_sha256 = decoded.record_sha256;
    var second_storage: [record.encoded_bytes]u8 = undefined;
    const second = try record.encodeV1(second_input, &second_storage);
    var stream: [record.encoded_bytes * 2]u8 = undefined;
    @memcpy(stream[0..record.encoded_bytes], encoded);
    @memcpy(stream[record.encoded_bytes..], second);
    const recovery_anchor: record.RecoveryAnchorV1 = .{
        .record_epoch = input.record_epoch,
        .next_sequence = 1,
        .previous_record_sha256 = capsule.zero_digest,
    };
    const clean_recovery = try record.classifyRecoveryV1(&stream, recovery_anchor);
    const short_recovery = try record.classifyRecoveryV1(
        stream[0 .. record.encoded_bytes + 100],
        recovery_anchor,
    );
    const body_recovery = try record.classifyRecoveryV1(
        stream[0 .. record.encoded_bytes + record.body_bytes],
        recovery_anchor,
    );
    const partial_footer_recovery = try record.classifyRecoveryV1(
        stream[0 .. record.encoded_bytes + record.body_bytes + 1],
        recovery_anchor,
    );
    var corrupt_stream = stream;
    corrupt_stream[record.encoded_bytes + record.accounting_before_offset] ^= 1;
    const corrupt_recovery = try record.classifyRecoveryV1(
        &corrupt_stream,
        recovery_anchor,
    );
    if (clean_recovery.status != .clean or
        clean_recovery.committed_records != 2 or
        short_recovery.status != .short_body_tail or
        body_recovery.status != .body_without_footer or
        partial_footer_recovery.status != .partial_footer_tail or
        corrupt_recovery.status != .corrupt_record or
        corrupt_recovery.committed_records != 1)
        return error.RecoveryClassificationMismatch;

    var writer_backing: [record.encoded_bytes * 2]u8 = undefined;
    var model_storage = try sweep_writer.DeterministicStorageV1.init(
        &writer_backing,
        encoded,
        41,
    );
    var writer_lease = try model_storage.acquire();
    var durable_writer = try sweep_writer.WriterV1.openClean(
        model_storage.bytes(),
        recovery_anchor,
        try writer_lease.appendCapability(),
    );
    const append_receipt = try durable_writer.appendRecord(second);
    if (!std.mem.eql(u8, model_storage.bytes(), &stream) or
        model_storage.trace().len != 4 or
        append_receipt.sequence != 2 or
        !append_receipt.body_sync_exercised or
        !append_receipt.footer_sync_exercised)
        return error.WriterContractMismatch;
    try writer_lease.release();

    var repair_backing: [record.encoded_bytes * 2]u8 = undefined;
    const repair_tail_bytes = record.body_bytes + 7;
    var repair_storage = try sweep_writer.DeterministicStorageV1.init(
        &repair_backing,
        stream[0 .. record.encoded_bytes + repair_tail_bytes],
        42,
    );
    var repair_lease = try repair_storage.acquire();
    const recovery_plan = try sweep_writer.planRecoveryV1(
        repair_storage.bytes(),
        recovery_anchor,
        repair_lease.snapshot,
    );
    var repairer = try sweep_writer.RepairerV1.init(
        repair_storage.bytes(),
        recovery_anchor,
        try repair_lease.prepareRepair(repair_storage.bytes(), recovery_anchor),
    );
    const repair_receipt = try repairer.apply();
    try repair_lease.release();
    var repaired_lease = try repair_storage.acquire();
    var repaired_writer = try sweep_writer.WriterV1.openClean(
        repair_storage.bytes(),
        recovery_anchor,
        try repaired_lease.appendCapability(),
    );
    _ = try repaired_writer.appendRecord(second);
    if (recovery_plan.action != .repair_incomplete_tail or
        recovery_plan.classification.status != .partial_footer_tail or
        repair_receipt.discarded_tail_bytes != repair_tail_bytes or
        !std.mem.eql(u8, repair_storage.bytes(), &stream))
        return error.RepairContractMismatch;
    try repaired_lease.release();

    var corrupted = storage;
    corrupted[record.accounting_before_offset] ^= 1;
    if (record.decodeV1(&corrupted)) |_| {
        return error.MutationWasAccepted;
    } else |_| {}

    const foreign_input = demoInput(0x7a, 0x7b);
    var foreign_storage: [record.encoded_bytes]u8 = undefined;
    const foreign = try record.encodeV1(foreign_input, &foreign_storage);
    _ = try record.decodeV1(foreign);
    if (record.decodeAndVerifyV1(foreign, expected)) |_| {
        return error.ForeignRecordWasAccepted;
    } else |err| switch (err) {
        record.Error.RecordExpectationMismatch => {},
        else => return err,
    }

    var encoded_sha256: record.Digest = undefined;
    std.crypto.hash.sha2.Sha256.hash(encoded, &encoded_sha256, .{});
    var stream_sha256: record.Digest = undefined;
    std.crypto.hash.sha2.Sha256.hash(&stream, &stream_sha256, .{});
    const record_hex = std.fmt.bytesToHex(decoded.record_sha256, .lower);
    const encoded_hex = std.fmt.bytesToHex(encoded_sha256, .lower);
    const sweep_commit_hex = std.fmt.bytesToHex(
        input.commit_receipt.commit_sha256,
        .lower,
    );
    const stream_hex = std.fmt.bytesToHex(stream_sha256, .lower);

    var stdout_buffer: [4096]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buffer);
    const stdout = &stdout_writer.interface;
    try stdout.print(
        "{{\"schema\":\"glacier.continuation-object-sweep-record/demo-v1\"," ++
            "\"encoded_bytes\":{d},\"body_bytes\":{d}," ++
            "\"commit_footer_bytes\":{d},\"sequence\":{d}," ++
            "\"previous_record_is_zero\":true," ++
            "\"entry_count_before\":{d},\"entry_count_after\":{d}," ++
            "\"retired_entries_before\":{d}," ++
            "\"retired_entries_after\":{d}," ++
            "\"payload_bytes_before\":{d}," ++
            "\"payload_bytes_after\":{d}," ++
            "\"freed_entries\":{d},\"freed_payload_bytes\":{d}," ++
            "\"freed_index_bytes\":{d}," ++
            "\"allocator_deallocation_calls\":{d}," ++
            "\"append_plan_body_first\":true," ++
            "\"mutation_rejected\":true," ++
            "\"valid_foreign_record_rejected\":true," ++
            "\"stream_records\":{d},\"stream_bytes\":{d}," ++
            "\"recovery_clean_status\":\"{s}\"," ++
            "\"recovery_short_status\":\"{s}\"," ++
            "\"recovery_body_status\":\"{s}\"," ++
            "\"recovery_partial_footer_status\":\"{s}\"," ++
            "\"recovery_corrupt_status\":\"{s}\"," ++
            "\"recovery_corrupt_safe_records\":{d}," ++
            "\"recovery_modifies_input\":false," ++
            "\"recovery_repair_authority\":false," ++
            "\"writer_snapshot_bound\":true," ++
            "\"writer_exclusive_lease\":true," ++
            "\"writer_phase_count\":{d}," ++
            "\"writer_body_sync\":true,\"writer_footer_sync\":true," ++
            "\"writer_poison_on_uncertain_io\":true," ++
            "\"repair_action\":\"{s}\",\"repair_status\":\"{s}\"," ++
            "\"repair_discarded_tail_bytes\":{d}," ++
            "\"repair_requires_reacquire\":true," ++
            "\"append_authority_can_truncate\":false," ++
            "\"repair_authority_can_append\":false," ++
            "\"heap_allocations\":0,\"filesystem_authority\":false," ++
            "\"network_authority\":false,\"clock_authority\":false," ++
            "\"deletion_authority\":false,\"recovery_authority\":false," ++
            "\"durable\":false," ++
            "\"record_sha256\":\"{s}\"," ++
            "\"encoded_sha256\":\"{s}\"," ++
            "\"stream_sha256\":\"{s}\"," ++
            "\"sweep_commit_sha256\":\"{s}\",\"verified\":true}}\n",
        .{
            encoded.len,
            plan.body.len,
            plan.commit_footer.len,
            input.sequence,
            input.store_receipt.accounting_before.entry_count,
            input.store_receipt.accounting_after.entry_count,
            input.store_receipt.accounting_before.retired_entries,
            input.store_receipt.accounting_after.retired_entries,
            input.store_receipt.accounting_before.payload_bytes,
            input.store_receipt.accounting_after.payload_bytes,
            input.commit_receipt.freed_entries,
            input.commit_receipt.freed_payload_bytes,
            input.commit_receipt.freed_index_bytes,
            input.commit_receipt.allocator_deallocation_calls,
            clean_recovery.committed_records,
            stream.len,
            @tagName(clean_recovery.status),
            @tagName(short_recovery.status),
            @tagName(body_recovery.status),
            @tagName(partial_footer_recovery.status),
            @tagName(corrupt_recovery.status),
            corrupt_recovery.committed_records,
            model_storage.trace().len,
            @tagName(recovery_plan.action),
            @tagName(recovery_plan.classification.status),
            repair_receipt.discarded_tail_bytes,
            &record_hex,
            &encoded_hex,
            &stream_hex,
            &sweep_commit_hex,
        },
    );
    try stdout.flush();
}

fn digest(byte: u8) record.Digest {
    return [_]u8{byte} ** @sizeOf(record.Digest);
}

fn demoInput(commit_challenge: u8, record_challenge: u8) record.InputV1 {
    const commit_grant: sweep.CommitGrantV1 = .{
        .authority_epoch = 11,
        .tenant_scope_sha256 = digest(0x61),
        .bundle_sha256 = digest(0x62),
        .store_grant_sha256 = digest(0x63),
        .sweep_grant_sha256 = digest(0x64),
        .prepare_sha256 = digest(0x65),
        .expected_snapshot_sha256 = digest(0x66),
        .collection_plan_sha256 = digest(0x67),
        .max_freed_entries = 2,
        .max_freed_bytes = 128,
        .challenge_sha256 = digest(commit_challenge),
    };
    const grant_sha256 = sweep.commitGrantRootV1(commit_grant) catch
        unreachable;
    var store_receipt: object_store.RetiredCommitReceiptV1 = .{
        .authorization_sha256 = grant_sha256,
        .targets_sha256 = digest(0x68),
        .snapshot_before_sha256 = commit_grant.expected_snapshot_sha256,
        .snapshot_after_sha256 = digest(0x69),
        .accounting_before = .{
            .entry_count = 8,
            .live_entries = 6,
            .quarantined_entries = 1,
            .retired_entries = 1,
            .payload_bytes = 255,
            .logical_index_bytes = 1024,
            .reference_count = 8,
            .active_leases = 1,
            .repair_count = 0,
        },
        .accounting_after = .{
            .entry_count = 7,
            .live_entries = 6,
            .quarantined_entries = 1,
            .retired_entries = 0,
            .payload_bytes = 216,
            .logical_index_bytes = 896,
            .reference_count = 8,
            .active_leases = 1,
            .repair_count = 0,
        },
        .freed_entries = 1,
        .freed_payload_bytes = 39,
        .freed_index_bytes = object_store.logical_index_entry_bytes,
        .freed_repair_count = 0,
        .allocator_deallocation_calls = 1,
        .commit_sha256 = capsule.zero_digest,
    };
    store_receipt.commit_sha256 =
        object_store.retiredCommitReceiptRootV1(store_receipt);
    var commit_receipt: sweep.CommitReceiptV1 = .{
        .commit_grant_sha256 = grant_sha256,
        .sweep_grant_sha256 = commit_grant.sweep_grant_sha256,
        .prepare_sha256 = commit_grant.prepare_sha256,
        .collection_plan_sha256 = commit_grant.collection_plan_sha256,
        .targets_sha256 = store_receipt.targets_sha256,
        .snapshot_before_sha256 = store_receipt.snapshot_before_sha256,
        .snapshot_after_sha256 = store_receipt.snapshot_after_sha256,
        .store_commit_sha256 = store_receipt.commit_sha256,
        .freed_entries = store_receipt.freed_entries,
        .freed_payload_bytes = store_receipt.freed_payload_bytes,
        .freed_index_bytes = store_receipt.freed_index_bytes,
        .freed_repair_count = store_receipt.freed_repair_count,
        .allocator_deallocation_calls = store_receipt.allocator_deallocation_calls,
        .commit_sha256 = capsule.zero_digest,
    };
    commit_receipt.commit_sha256 = sweep.commitRootV1(commit_receipt);
    return .{
        .record_epoch = 0x5357_4545_5000_0001,
        .sequence = 1,
        .previous_record_sha256 = capsule.zero_digest,
        .record_challenge_sha256 = digest(record_challenge),
        .commit_grant = commit_grant,
        .commit_receipt = commit_receipt,
        .store_receipt = store_receipt,
    };
}

File: bench/continuation_object_sweep_record.py
"""Independent codec and verifier for continuation sweep evidence records."""

from __future__ import annotations

import hashlib
import struct
from typing import Any

from bench import continuation_object_store as object_store
from bench import continuation_object_sweep as sweep


class SweepRecordError(ValueError):
    """The record framing, chain, expectation, or embedded evidence is invalid."""


Record = dict[str, Any]
ZERO_DIGEST = bytes(32)
ABI_VERSION = 0x4743_5352_0000_0001
BODY_MAGIC = b"GCSWRB01"
COMMIT_MAGIC = b"GCSWRF01"
ALLOWED_FLAGS = 0
RECORD_DOMAIN = b"glacier-continuation-sweep-record-body-v1\x00"
ACCOUNTING_FIELDS = (
    "entry_count",
    "live_entries",
    "quarantined_entries",
    "retired_entries",
    "payload_bytes",
    "logical_index_bytes",
    "reference_count",
    "active_leases",
    "repair_count",
)
FREED_FIELDS = (
    "freed_entries",
    "freed_payload_bytes",
    "freed_index_bytes",
    "freed_repair_count",
    "allocator_deallocation_calls",
)
ACCOUNTING_BEFORE_OFFSET = 456
SEQUENCE_OFFSET = 40
BODY_PREFIX_BYTES = 704
BODY_BYTES = 736
COMMIT_FOOTER_BYTES = 48
ENCODED_BYTES = 784


def _u32(value: int) -> bytes:
    if not isinstance(value, int) or not 0 <= value <= 0xFFFFFFFF:
        raise SweepRecordError("u32 out of range")
    return struct.pack("<I", value)


def _u64(value: int) -> bytes:
    if not isinstance(value, int) or not 0 <= value <= 0xFFFFFFFFFFFFFFFF:
        raise SweepRecordError("u64 out of range")
    return struct.pack("<Q", value)


def _nonzero_digest(value: bytes) -> bytes:
    if not isinstance(value, bytes) or len(value) != 32 or value == ZERO_DIGEST:
        raise SweepRecordError("invalid digest")
    return value


def _chain_digest(value: bytes) -> bytes:
    if not isinstance(value, bytes) or len(value) != 32:
        raise SweepRecordError("invalid chain digest")
    return value


def _accounting_bytes(accounting: Record) -> bytes:
    try:
        return b"".join(_u64(accounting[name]) for name in ACCOUNTING_FIELDS)
    except KeyError as exc:
        raise SweepRecordError("missing accounting field") from exc


def record_root(prefix: bytes) -> bytes:
    if not isinstance(prefix, bytes) or len(prefix) != BODY_PREFIX_BYTES:
        raise SweepRecordError("invalid record prefix")
    return hashlib.sha256(RECORD_DOMAIN + prefix).digest()


def _validate_input(record: Record) -> None:
    try:
        record_epoch = record["record_epoch"]
        sequence = record["sequence"]
        previous = _chain_digest(record["previous_record_sha256"])
        _nonzero_digest(record["record_challenge_sha256"])
        commit_grant = record["commit_grant"]
        commit_receipt = record["commit_receipt"]
        store_receipt = record["store_receipt"]
        _u64(record_epoch)
        _u64(sequence)
    except KeyError as exc:
        raise SweepRecordError("missing sweep record field") from exc
    if record_epoch == 0 or sequence == 0:
        raise SweepRecordError("invalid record chain position")
    if (sequence == 1) != (previous == ZERO_DIGEST):
        raise SweepRecordError("invalid previous record root")
    try:
        sweep.verify_commit_receipt(
            commit_grant,
            commit_receipt,
            store_receipt,
        )
    except (KeyError, TypeError, sweep.SweepError) as exc:
        raise SweepRecordError("invalid embedded sweep evidence") from exc


def encode(record: Record) -> bytes:
    _validate_input(record)
    grant = record["commit_grant"]
    receipt = record["commit_receipt"]
    store_receipt = record["store_receipt"]
    try:
        parts = [
            BODY_MAGIC,
            _u64(ABI_VERSION),
            _u64(ENCODED_BYTES),
            _u32(ALLOWED_FLAGS),
            _u32(0),
            _u64(record["record_epoch"]),
            _u64(record["sequence"]),
            _chain_digest(record["previous_record_sha256"]),
            _nonzero_digest(record["record_challenge_sha256"]),
            _u64(grant["authority_epoch"]),
            _nonzero_digest(grant["tenant_scope_sha256"]),
            _nonzero_digest(grant["bundle_sha256"]),
            _nonzero_digest(grant["store_grant_sha256"]),
            _nonzero_digest(grant["sweep_grant_sha256"]),
            _nonzero_digest(grant["prepare_sha256"]),
            _nonzero_digest(grant["expected_snapshot_sha256"]),
            _nonzero_digest(grant["collection_plan_sha256"]),
            _u64(grant["max_freed_entries"]),
            _u64(grant["max_freed_bytes"]),
            _nonzero_digest(grant["challenge_sha256"]),
            _nonzero_digest(receipt["targets_sha256"]),
            _nonzero_digest(receipt["snapshot_after_sha256"]),
            _accounting_bytes(store_receipt["accounting_before"]),
            _accounting_bytes(store_receipt["accounting_after"]),
            *(_u64(receipt[name]) for name in FREED_FIELDS),
            _nonzero_digest(store_receipt["commit_sha256"]),
            _nonzero_digest(receipt["commit_sha256"]),
        ]
    except KeyError as exc:
        raise SweepRecordError("missing embedded evidence field") from exc
    prefix = b"".join(parts)
    if len(prefix) != BODY_PREFIX_BYTES:
        raise SweepRecordError("record layout drift")
    root = record_root(prefix)
    encoded = b"".join(
        (
            prefix,
            root,
            COMMIT_MAGIC,
            _u64(record["sequence"]),
            root,
        )
    )
    if len(encoded) != ENCODED_BYTES:
        raise SweepRecordError("record length drift")
    return encoded


class _Reader:
    def __init__(self, encoded: bytes) -> None:
        self.encoded = encoded
        self.position = 0

    def take(self, length: int) -> bytes:
        end = self.position + length
        if length < 0 or end > len(self.encoded):
            raise SweepRecordError("truncated record")
        value = self.encoded[self.position : end]
        self.position = end
        return value

    def u32(self) -> int:
        return struct.unpack("<I", self.take(4))[0]

    def u64(self) -> int:
        return struct.unpack("<Q", self.take(8))[0]

    def digest(self) -> bytes:
        return self.take(32)

    def accounting(self) -> Record:
        return {name: self.u64() for name in ACCOUNTING_FIELDS}


def _decode(encoded: bytes) -> Record:
    if not isinstance(encoded, bytes) or len(encoded) != ENCODED_BYTES:
        raise SweepRecordError("invalid encoded length")
    reader = _Reader(encoded)
    if reader.take(len(BODY_MAGIC)) != BODY_MAGIC:
        raise SweepRecordError("invalid body magic")
    if reader.u64() != ABI_VERSION:
        raise SweepRecordError("invalid ABI")
    if reader.u64() != ENCODED_BYTES:
        raise SweepRecordError("invalid declared length")
    if reader.u32() != ALLOWED_FLAGS or reader.u32() != 0:
        raise SweepRecordError("invalid flags")

    record_epoch = reader.u64()
    sequence = reader.u64()
    previous_record_sha256 = reader.digest()
    record_challenge_sha256 = reader.digest()
    grant = {
        "authority_epoch": reader.u64(),
        "tenant_scope_sha256": reader.digest(),
        "bundle_sha256": reader.digest(),
        "store_grant_sha256": reader.digest(),
        "sweep_grant_sha256": reader.digest(),
        "prepare_sha256": reader.digest(),
        "expected_snapshot_sha256": reader.digest(),
        "collection_plan_sha256": reader.digest(),
        "max_freed_entries": reader.u64(),
        "max_freed_bytes": reader.u64(),
        "challenge_sha256": reader.digest(),
    }
    targets_sha256 = reader.digest()
    snapshot_after_sha256 = reader.digest()
    accounting_before = reader.accounting()
    accounting_after = reader.accounting()
    freed = {name: reader.u64() for name in FREED_FIELDS}
    store_commit_sha256 = reader.digest()
    sweep_commit_sha256 = reader.digest()
    if reader.position != BODY_PREFIX_BYTES:
        raise SweepRecordError("invalid body prefix length")

    stored_record_sha256 = reader.digest()
    computed_record_sha256 = record_root(encoded[:BODY_PREFIX_BYTES])
    if stored_record_sha256 != computed_record_sha256:
        raise SweepRecordError("body root mismatch")
    if reader.position != BODY_BYTES:
        raise SweepRecordError("invalid body length")
    if reader.take(len(COMMIT_MAGIC)) != COMMIT_MAGIC:
        raise SweepRecordError("invalid commit footer")
    if reader.u64() != sequence:
        raise SweepRecordError("commit sequence mismatch")
    if reader.digest() != stored_record_sha256 or reader.position != len(encoded):
        raise SweepRecordError("commit root mismatch")

    try:
        commit_grant_sha256 = sweep.commit_grant_root(grant)
    except (KeyError, TypeError, sweep.SweepError) as exc:
        raise SweepRecordError("invalid commit grant") from exc
    store_receipt = {
        "authorization_sha256": commit_grant_sha256,
        "targets_sha256": targets_sha256,
        "snapshot_before_sha256": grant["expected_snapshot_sha256"],
        "snapshot_after_sha256": snapshot_after_sha256,
        "accounting_before": accounting_before,
        "accounting_after": accounting_after,
        **freed,
        "commit_sha256": store_commit_sha256,
    }
    commit_receipt = {
        "commit_grant_sha256": commit_grant_sha256,
        "sweep_grant_sha256": grant["sweep_grant_sha256"],
        "prepare_sha256": grant["prepare_sha256"],
        "collection_plan_sha256": grant["collection_plan_sha256"],
        "targets_sha256": targets_sha256,
        "snapshot_before_sha256": grant["expected_snapshot_sha256"],
        "snapshot_after_sha256": snapshot_after_sha256,
        "store_commit_sha256": store_commit_sha256,
        **freed,
        "commit_sha256": sweep_commit_sha256,
    }
    record = {
        "record_epoch": record_epoch,
        "sequence": sequence,
        "previous_record_sha256": previous_record_sha256,
        "record_challenge_sha256": record_challenge_sha256,
        "commit_grant": grant,
        "commit_receipt": commit_receipt,
        "store_receipt": store_receipt,
    }
    _validate_input(record)
    return {"input": record, "record_sha256": stored_record_sha256}


def decode(encoded: bytes) -> Record:
    try:
        return _decode(encoded)
    except SweepRecordError:
        raise
    except (KeyError, TypeError, ValueError, struct.error) as exc:
        raise SweepRecordError("invalid sweep record") from exc


def expectation(record: Record, record_sha256: bytes) -> Record:
    _validate_input(record)
    expected = {
        "record_epoch": record["record_epoch"],
        "sequence": record["sequence"],
        "previous_record_sha256": record["previous_record_sha256"],
        "sweep_commit_sha256": record["commit_receipt"]["commit_sha256"],
        "record_sha256": _nonzero_digest(record_sha256),
    }
    _validate_expectation(expected)
    return expected


def _validate_expectation(expected: Record) -> None:
    try:
        record_epoch = expected["record_epoch"]
        sequence = expected["sequence"]
        previous = _chain_digest(expected["previous_record_sha256"])
        _nonzero_digest(expected["sweep_commit_sha256"])
        _nonzero_digest(expected["record_sha256"])
        _u64(record_epoch)
        _u64(sequence)
    except KeyError as exc:
        raise SweepRecordError("missing expectation field") from exc
    if record_epoch == 0 or sequence == 0:
        raise SweepRecordError("invalid expected chain position")
    if (sequence == 1) != (previous == ZERO_DIGEST):
        raise SweepRecordError("invalid expected previous root")


def decode_and_verify(encoded: bytes, expected: Record) -> Record:
    _validate_expectation(expected)
    decoded = decode(encoded)
    record = decoded["input"]
    if (
        record["record_epoch"] != expected["record_epoch"]
        or record["sequence"] != expected["sequence"]
        or record["previous_record_sha256"]
        != expected["previous_record_sha256"]
        or record["commit_receipt"]["commit_sha256"]
        != expected["sweep_commit_sha256"]
        or decoded["record_sha256"] != expected["record_sha256"]
    ):
        raise SweepRecordError("record expectation mismatch")
    return decoded


def append_plan(encoded: bytes) -> Record:
    decoded = decode(encoded)
    return {
        "body": encoded[:BODY_BYTES],
        "commit_footer": encoded[BODY_BYTES:],
        "record_sha256": decoded["record_sha256"],
    }


def origin_recovery_anchor(record_epoch: int = 0x5357_4545_5000_0001) -> Record:
    anchor = {
        "record_epoch": record_epoch,
        "next_sequence": 1,
        "previous_record_sha256": ZERO_DIGEST,
    }
    _validate_recovery_anchor(anchor)
    return anchor


def _validate_recovery_anchor(anchor: Record) -> None:
    try:
        record_epoch = anchor["record_epoch"]
        next_sequence = anchor["next_sequence"]
        previous = _chain_digest(anchor["previous_record_sha256"])
        _u64(record_epoch)
        _u64(next_sequence)
    except KeyError as exc:
        raise SweepRecordError("missing recovery anchor field") from exc
    if record_epoch == 0 or next_sequence == 0:
        raise SweepRecordError("invalid recovery anchor position")
    if (next_sequence == 1) != (previous == ZERO_DIGEST):
        raise SweepRecordError("invalid recovery anchor chain")


def _expected_footer(body: bytes) -> bytes:
    if not isinstance(body, bytes) or len(body) != BODY_BYTES:
        raise SweepRecordError("invalid recovery body")
    return b"".join(
        (
            COMMIT_MAGIC,
            body[SEQUENCE_OFFSET : SEQUENCE_OFFSET + 8],
            body[BODY_PREFIX_BYTES:BODY_BYTES],
        )
    )


def _decode_body_for_recovery(body: bytes) -> Record:
    return decode(body + _expected_footer(body))


def _recovery_classification(
    status: str,
    anchor: Record,
    last_sequence: int,
    committed_records: int,
    committed_bytes: int,
    tail_bytes: int,
    final_record_sha256: bytes,
) -> Record:
    return {
        "status": status,
        "record_epoch": anchor["record_epoch"],
        "first_sequence": anchor["next_sequence"],
        "last_sequence": last_sequence,
        "committed_records": committed_records,
        "committed_bytes": committed_bytes,
        "tail_bytes": tail_bytes,
        "final_record_sha256": final_record_sha256,
    }


def _recovery_record_matches(
    decoded: Record,
    record_epoch: int,
    sequence: int,
    previous_record_sha256: bytes,
) -> bool:
    value = decoded["input"]
    return (
        value["record_epoch"] == record_epoch
        and value["sequence"] == sequence
        and value["previous_record_sha256"] == previous_record_sha256
    )


def classify_recovery(stream: bytes, anchor: Record) -> Record:
    """Classify a record stream without modifying it or granting repair authority."""

    if not isinstance(stream, bytes):
        raise SweepRecordError("invalid recovery stream")
    _validate_recovery_anchor(anchor)
    offset = 0
    committed_records = 0
    expected_sequence = anchor["next_sequence"]
    expected_previous = anchor["previous_record_sha256"]
    last_sequence = expected_sequence - 1
    final_record_sha256 = expected_previous

    while len(stream) - offset >= ENCODED_BYTES:
        try:
            decoded = decode(stream[offset : offset + ENCODED_BYTES])
        except SweepRecordError:
            return _recovery_classification(
                "corrupt_record",
                anchor,
                last_sequence,
                committed_records,
                offset,
                len(stream) - offset,
                final_record_sha256,
            )
        if not _recovery_record_matches(
            decoded,
            anchor["record_epoch"],
            expected_sequence,
            expected_previous,
        ):
            return _recovery_classification(
                "corrupt_record",
                anchor,
                last_sequence,
                committed_records,
                offset,
                len(stream) - offset,
                final_record_sha256,
            )
        offset += ENCODED_BYTES
        committed_records += 1
        last_sequence = decoded["input"]["sequence"]
        final_record_sha256 = decoded["record_sha256"]
        expected_previous = final_record_sha256
        if offset < len(stream):
            if last_sequence == 0xFFFFFFFFFFFFFFFF:
                return _recovery_classification(
                    "corrupt_record",
                    anchor,
                    last_sequence,
                    committed_records,
                    offset,
                    len(stream) - offset,
                    final_record_sha256,
                )
            expected_sequence = last_sequence + 1

    tail_bytes = len(stream) - offset
    if tail_bytes == 0:
        return _recovery_classification(
            "clean",
            anchor,
            last_sequence,
            committed_records,
            offset,
            0,
            final_record_sha256,
        )
    if tail_bytes < BODY_BYTES:
        return _recovery_classification(
            "short_body_tail",
            anchor,
            last_sequence,
            committed_records,
            offset,
            tail_bytes,
            final_record_sha256,
        )

    body = stream[offset : offset + BODY_BYTES]
    try:
        decoded_body = _decode_body_for_recovery(body)
    except SweepRecordError:
        return _recovery_classification(
            "corrupt_record",
            anchor,
            last_sequence,
            committed_records,
            offset,
            tail_bytes,
            final_record_sha256,
        )
    if not _recovery_record_matches(
        decoded_body,
        anchor["record_epoch"],
        expected_sequence,
        expected_previous,
    ):
        return _recovery_classification(
            "corrupt_record",
            anchor,
            last_sequence,
            committed_records,
            offset,
            tail_bytes,
            final_record_sha256,
        )
    if tail_bytes == BODY_BYTES:
        return _recovery_classification(
            "body_without_footer",
            anchor,
            last_sequence,
            committed_records,
            offset,
            tail_bytes,
            final_record_sha256,
        )

    partial_footer = stream[offset + BODY_BYTES :]
    if partial_footer != _expected_footer(body)[: len(partial_footer)]:
        return _recovery_classification(
            "corrupt_record",
            anchor,
            last_sequence,
            committed_records,
            offset,
            tail_bytes,
            final_record_sha256,
        )
    return _recovery_classification(
        "partial_footer_tail",
        anchor,
        last_sequence,
        committed_records,
        offset,
        tail_bytes,
        final_record_sha256,
    )


def demo_input(commit_challenge: int = 0x6A, record_challenge: int = 0x6B) -> Record:
    def digest(byte: int) -> bytes:
        return bytes((byte,)) * 32

    grant = {
        "authority_epoch": 11,
        "tenant_scope_sha256": digest(0x61),
        "bundle_sha256": digest(0x62),
        "store_grant_sha256": digest(0x63),
        "sweep_grant_sha256": digest(0x64),
        "prepare_sha256": digest(0x65),
        "expected_snapshot_sha256": digest(0x66),
        "collection_plan_sha256": digest(0x67),
        "max_freed_entries": 2,
        "max_freed_bytes": 128,
        "challenge_sha256": digest(commit_challenge),
    }
    grant_sha256 = sweep.commit_grant_root(grant)
    store_receipt = {
        "authorization_sha256": grant_sha256,
        "targets_sha256": digest(0x68),
        "snapshot_before_sha256": grant["expected_snapshot_sha256"],
        "snapshot_after_sha256": digest(0x69),
        "accounting_before": {
            "entry_count": 8,
            "live_entries": 6,
            "quarantined_entries": 1,
            "retired_entries": 1,
            "payload_bytes": 255,
            "logical_index_bytes": 1024,
            "reference_count": 8,
            "active_leases": 1,
            "repair_count": 0,
        },
        "accounting_after": {
            "entry_count": 7,
            "live_entries": 6,
            "quarantined_entries": 1,
            "retired_entries": 0,
            "payload_bytes": 216,
            "logical_index_bytes": 896,
            "reference_count": 8,
            "active_leases": 1,
            "repair_count": 0,
        },
        "freed_entries": 1,
        "freed_payload_bytes": 39,
        "freed_index_bytes": 128,
        "freed_repair_count": 0,
        "allocator_deallocation_calls": 1,
    }
    store_receipt["commit_sha256"] = object_store.retired_commit_receipt_root(
        store_receipt
    )
    receipt = {
        "commit_grant_sha256": grant_sha256,
        "sweep_grant_sha256": grant["sweep_grant_sha256"],
        "prepare_sha256": grant["prepare_sha256"],
        "collection_plan_sha256": grant["collection_plan_sha256"],
        "targets_sha256": store_receipt["targets_sha256"],
        "snapshot_before_sha256": store_receipt["snapshot_before_sha256"],
        "snapshot_after_sha256": store_receipt["snapshot_after_sha256"],
        "store_commit_sha256": store_receipt["commit_sha256"],
        "freed_entries": 1,
        "freed_payload_bytes": 39,
        "freed_index_bytes": 128,
        "freed_repair_count": 0,
        "allocator_deallocation_calls": 1,
    }
    receipt["commit_sha256"] = sweep.commit_root(receipt)
    return {
        "record_epoch": 0x5357_4545_5000_0001,
        "sequence": 1,
        "previous_record_sha256": ZERO_DIGEST,
        "record_challenge_sha256": digest(record_challenge),
        "commit_grant": grant,
        "commit_receipt": receipt,
        "store_receipt": store_receipt,
    }
